session/schema_amender.go

```diff
@@ -0,0 +1,532 @@
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend failed to get snapshot using forUpdateTS", zap.Error(err))
+		return nil, errors.Trace(err)
+	}
+	oldValKvMap, err := snapshot.BatchGet(ctx, removeKeys)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend failed to batch get kv old keys", zap.Error(err))
+		return nil, errors.Trace(err)
+	}
+
+	res := &rowKvMap{
+		oldRowKvMap: oldValKvMap,
+		newRowKvMap: newValKvMap,
+	}
+	return res, nil
+}
+
+// genAllAmendMutations generates CommitterMutations for all tables and related amend operations.
+func (s *SchemaAmender) genAllAmendMutations(ctx context.Context, commitMutations tikv.CommitterMutations,
+	info *amendCollector) (*tikv.CommitterMutations, error) {
+	rowKvMap, err := s.prepareKvMap(ctx, commitMutations, info)
+	if err != nil {
+		return nil, err
+	}
+	// Do generate add/remove mutations processing each key.
+	resultNewMutations := tikv.NewCommiterMutations(32)
+	for _, amendOps := range info.tblAmendOpMap {
+		for _, curOp := range amendOps {
+			err := curOp.genMutations(ctx, s.sess, commitMutations, rowKvMap, &resultNewMutations)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return &resultNewMutations, nil
+}
+
+// AmendTxn does check and generate amend mutations based on input infoSchema and mutations, mutations need to prewrite
+// are returned, the input commitMutations will not be changed.
+func (s *SchemaAmender) AmendTxn(ctx context.Context, startInfoSchema tikv.SchemaVer, change *tikv.RelatedSchemaChange,
+	commitMutations tikv.CommitterMutations) (*tikv.CommitterMutations, error) {
+	// Get info schema meta
+	infoSchemaAtStart := startInfoSchema.(infoschema.InfoSchema)
+	infoSchemaAtCheck := change.LatestInfoSchema.(infoschema.InfoSchema)
+
+	// Collect amend operations for each table by physical table ID.
+	var needAmendMem bool
+	amendCollector := newAmendCollector()
+	for i, tblID := range change.PhyTblIDS {
+		actionType := change.ActionTypes[i]
+		// Check amendable flags, return if not supported flags exist.
+		if actionType&(^amendableType) != 0 {
+			logutil.Logger(ctx).Info("amend action type not supported for txn", zap.Int64("tblID", tblID), zap.Uint64("actionType", actionType))
+			return nil, errors.Trace(table.ErrUnsupportedOp)
+		}
+		// Partition table is not supported now.
+		tblInfoAtStart, ok := infoSchemaAtStart.TableByID(tblID)
+		if !ok {
+			return nil, errors.Trace(errors.Errorf("tableID=%d is not found in infoSchema", tblID))
+		}
+		if tblInfoAtStart.Meta().Partition != nil {
+			logutil.Logger(ctx).Info("Amend for partition table is not supported",
+				zap.String("tableName", tblInfoAtStart.Meta().Name.String()), zap.Int64("tableID", tblID))
+			return nil, errors.Trace(table.ErrUnsupportedOp)
+		}
+		tblInfoAtCommit, ok := infoSchemaAtCheck.TableByID(tblID)
+		if !ok {
+			return nil, errors.Trace(errors.Errorf("tableID=%d is not found in infoSchema", tblID))
+		}
+		if actionType&(memBufAmendType) != 0 {
```

Comment:
If this `actionType` include `add index` and other DDL operations, do we need special treatment？

Reply:
In `collectTblAmendOps` we need to check for each possible memory amend type, it's like
```
func collectTblAmendOps {
   if actionType contains addIndexType {
       collect amend operations for index needs to amend
   }
   if actionType contains otherTypes {
       collect amend operations for this amend type
   }
}
```
Currently, only `add index` is considered for transaction memory amend.

Reply:
Got it.